ch4: misc clean up

hzhou opened the pull request:
## Pull Request Description

Cherry picking some misc patches from PR #4385. They are necessary fixes before we remove VCI Global lock.

## Expected Impact

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Add Devel Docs in the `doc/` directory for any new code design

hzhou commented:
test:mpich/warnings/quick

hzhou commented:
test:mpich/async

hzhou commented:
test:mpich/async

hzhou commented:
test:mpich/warnins
test:mpich/async

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/warnings

raffenet commented on src/mpid/ch4/netmod/ofi/ofi_init.c:
```diff
@@ -741,9 +741,8 @@ int MPIDI_OFI_mpi_finalize_hook(void)
-    mpi_errno =
-        MPIR_Allreduce(&barrier[0], &barrier[1], 1, MPI_INT, MPI_SUM, MPIR_Process.comm_world,
-                       &errflag);
+    mpi_errno = MPIR_Allreduce_allcomm_auto(&barrier[0], &barrier[1], 1, MPI_INT, MPI_SUM,
```
Does this override the user environment setting? Just trying to understand the scenario.

hzhou replied:
I think the auto method just runs the default json selection, skipping user choice.

hzhou commented:
Hi @raffenet , I removed the ofi `MPIR_Allreduce` commit from this PR. I believe at some point I hit test failures due to it tries to run non-blocking algorithm, however I couldn't recall at which point. We can delay this commit until necessary.

I also modified the `contextid` by completely commenting out the `MPIR_Allreduce`. It tries to run a blocking collective inside a progress loop of a non-blocking context id allocation, which really complicates things up. It appears all it does is just to get better error message, which I don't think it worth the complication. I am sure once we sit down and study the algorithm, there will be better solutions to achieve the same thing. For now, I suggest to simplify it. -- I can't remember how many times I have been patching up code due to this complication by now.

hzhou commented:
test:mpich/warnings
test:mpich/async

hzhou commented:
test:mpich/warnings
test:mpich/async
